### PR TITLE
topology: Add CML topology support for da7219+max98357a variant

### PR DESCRIPTION
**Pull request by macchian**
add the topology files for Cometlake
with Intel ASoC DMIC capture, HDMI playback
with analog codec da7219 on SSP0 for headphone playback/headset mic
with digital codec max98357a on SSP1 for speaker playback

Signed-off-by: Mac Chiang <mac.chiang@intel.com>

### CODE REVIEW

**Comment by ranj063** [comment]
please squash the two commits and fix the commit message with proper prefix: "topology: cml: ...."

**Comment by macchian** [comment]
> In addition to what @ranj063 wrote, please add also Signed-off-by line to every commit.

Thanks, @ranj063, @jajanusz . I can't find a way to modify commit messages alone in these two commits? Should I delete them and resubmit again?

**Comment by jajanusz** [comment]
@macchian Just do changes in your git repo and force push them. F.e.:
```
git rebase --interactive HEAD~2
# then in editor squash 2nd commit
# then rebase will continue and will ask your for commit message for both, update it
git push --force
```

**Comment by macchian** [comment]
> @macchian Just do changes in your git repo and force push them. F.e.:
> 
> ```
> git rebase --interactive HEAD~2
> # then in editor squash 2nd commit
> # then rebase will continue and will ask your for commit message for both, update it
> git push --force
> ```

Thanks for your guidance. updated for your reviewing!

**Comment by plbossart** [comment]
@macchian is there a machine driver for this configuration?

**Comment by ranj063** [comment]
@macchian any particular reason to split this topology into 2 m4 files? Do you anticipate to have other variants with sof-cml-da7219.m4 as the base and a different codec for the speaker? 

Also please squash the 2 commits.

**Comment by macchian** [comment]
> @macchian is there a machine driver for this configuration?

@plbossart , yes, I had it and verified. I will upstream soon for reviewing.

**Comment by macchian** [comment]
> @macchian any particular reason to split this topology into 2 m4 files? Do you anticipate to have other variants with sof-cml-da7219.m4 as the base and a different codec for the speaker?
> 
> Also please squash the 2 commits.

so far no. sof-cml-da7219 just pre-reserved. Because we might have setupbox sku in the feature. Any suggestion?

**Comment by plbossart** [comment]
> > @macchian is there a machine driver for this configuration?
> 
> @plbossart , yes, I had it and verified. I will upstream soon for reviewing.

I don't want a new machine driver for this configuration which looks very similar to APL/GLK with the same parts. We need to reuse and configure the links. Please submit your code asap.

File: tools/topology/sof-cml-da7219.m4 [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+	SSP_CONFIG(I2S, SSP_CLOCK(mclk, 24000000, codec_mclk_in),
+		      SSP_CLOCK(bclk, 2400000, codec_slave),
+		      SSP_CLOCK(fsync, 48000, codec_slave),
+		      SSP_TDM(2, 25, 3, 3),
+		      SSP_CONFIG_DATA(SSP, 0, 16)))
+
+# dmic01 (ID: 2)
+DAI_CONFIG(DMIC, 0, 2, dmic01,
+	   DMIC_CONFIG(1, 500000, 4800000, 40, 60, 48000,
+		DMIC_WORD_LENGTH(s32le), DMIC, 0,
```
Comment (xiulipan) [inline tools/topology/sof-cml-da7219.m4]:
```suggestion
		DMIC_WORD_LENGTH(s32le), 400, DMIC, 0,
```

Please check https://github.com/thesofproject/sof/pull/1515 (909b790235cba6cfa5ef6da9f9e2712ef4734d75) for this DMIC setting change.

**Comment by macchian** [comment]
> > > @macchian is there a machine driver for this configuration?
> > 
> > 
> > @plbossart , yes, I had it and verified. I will upstream soon for reviewing.
> 
> I don't want a new machine driver for this configuration which looks very similar to APL/GLK with the same parts. We need to reuse and configure the links. Please submit your code asap.

@plbossart , thanks for advises. Yes, I reuse the kbl machine driver for cml platform instead. submit here for review https://patchwork.kernel.org/patch/11070247/

File: tools/topology/sof-cml-da7219.m4 (hunk 1 above)
Comment (macchian) [reply tools/topology/sof-cml-da7219.m4]: Thanks @xiulipan , I added it and updated the new tplg for reviewing.  https://github.com/thesofproject/sof/commit/0115328d63317769bc712a96d13aa0c2011fda71

**Comment by macchian** [comment]
@xiulipan @ranj063 @jajanusz ,
  I'm doing this commit rebase into master branch. However, the file uncounted a build error as below:
         "fatal error: Invalid parameters (9) to PIPELINE_PCM_ADD"

  Per checking the patch list, I found the dai parameters increased by following patch
         "topology: enable pcm range and pipeline rate and remove frame count"

   I modified all these relevant changes in this commit
example:
 PIPELINE_PCM_ADD(sof/pipe-volume-playback.m4,
        1, 1, 2, s32le,
-       48, 1000, 0, 0)
+       1000, 0, 0,
+       48000, 48000, 48000)

  When applied this modified tplg, I can not playback as usual on device. I saw the errors 

[    8.741126] sof-audio-pci 0000:00:1f.3: error: no reply expected, received 0x0
[    8.830050] sof-audio-pci 0000:00:1f.3: firmware boot complete
[   12.729620] sof-audio-pci 0000:00:1f.3: error: no reply expected, received 0x0
[   12.818602] sof-audio-pci 0000:00:1f.3: firmware boot complete
[   12.844939] sof-audio-pci 0000:00:1f.3: error: ipc error for 0x60010000 size 20
[   12.844950] sof-audio-pci 0000:00:1f.3: error: hw params ipc failed for stream 1
[   12.844955] sof-audio-pci 0000:00:1f.3: ASoC: 0000:00:1f.3 hw params failed: -22
[   12.844960]  Speakers: ASoC: hw_params FE failed -22

  Do I miss anythings from the commits?

commit 77b6db88948f435114991f9b6e782dc56c983010
Author: Jaska Uimonen <jaska.uimonen@intel.com>
Date:   Thu Jan 3 14:11:21 2019 +0200

         "topology: enable pcm range and pipeline rate and remove frame count"

**Comment by xiulipan** [comment]
@macchian 
As there are some tplg and fw chagne. Some new TPGL will not work with old FW.

You can see that https://github.com/thesofproject/sof/pull/842 have a ton of fw patches also. So if you are working with some release branch fw, you should not use the tplg from master.

**Comment by macchian** [comment]
> @macchian
> As there are some tplg and fw chagne. Some new TPGL will not work with old FW.
> 
> You can see that #842 have a ton of fw patches also. So if you are working with some release branch fw, you should not use the tplg from master.

@xiulipan , thanks for your information.  Currently I'm using the remotes/origin/cml-004-drop-stable and remotes/origin/cml-005-drop-stable fw image. And the tplg issue is existing on both versions. Do you have suggested FW branch to test?

In the meanwhile, do you think the commit can be approved to merge in master with specific FW branch version?

**Comment by xiulipan** [comment]
@macchian Then you need to rebase this patch to the FW branch cml-004-drop-stable or cml-005-drop-stable.

The key point is to make tplg binary build out from the same FW SOF repo

**Comment by macchian** [comment]
> The key point is to make tplg binary build out from the same FW SOF repo

Thanks your advices! @xiulipan 
I created a new PR to rebase to specific fw branch. The link is #1808 for your reviewing.
For #1662, I will modify it working on latest fw version.

**Comment by tlauda** [comment]
@macchian Should we close this one?

**Comment by macchian** [comment]
> @macchian Should we close this one?

yes, we can close it. I've submitted the new version PR #1894 base on SOF 0.7 release.

**Pull request closed without merging**